### PR TITLE
Paul/coinrank

### PR DESCRIPTION
**Pull request by paullinator**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203527749920640

### CODE REVIEW

File: src/exchangeRateRouter.ts [hunk 1]
```diff
@@ -301,6 +310,108 @@ const sendExchangeRates: express.RequestHandler = (req, res, next): void => {
...
+          price: m.price * rate,
+          marketCap: m.marketCap * rate,
+          volume24h: m.volume24h * rate
+        }))
+
+        res.json({ data })
+
+        // Update redis cache
+        const redisData: CoinrankRedis = {
+          markets: data,
```
Comment (peachbits) [inline src/exchangeRateRouter.ts:390]: This only saves the the sliced data for the requested fiat. Convert everything in the USD list to the selected currency before saving it and returning to the user

File: src/exchangeRateRouter.ts [hunk 2]
```diff
@@ -301,6 +310,108 @@ const sendExchangeRates: express.RequestHandler = (req, res, next): void => {
...
+        .send(`Invalid start param: ${start}. Must be between 1-2000`)
+      return
+    }
+    if (length < 1 || length > 100) {
+      res
+        .status(400)
+        .send(`Invalid length param: ${length}. Must be between 1-100`)
+      return
+    }
+    const fiatSuffix = fiatCode.split(':')[1]
```
Comment (peachbits) [inline src/exchangeRateRouter.ts:349]: Nitpick, just use the fiatCode as is for the key. And replace any hardcoded USD/iso:USD value with the default fiat from the config

**Review by peachbits — CHANGES_REQUESTED** [review]

File: src/coinrankEngine.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import fetch from 'node-fetch'
+import { createClient } from 'redis'
+
+import { config } from './config'
+import { REDIS_COINRANK_KEY_PREFIX } from './constants'
+import { asCoingeckoMarkets, CoinrankRedis } from './types'
+import { setAsync, slackMessage } from './utils/dbUtils'
+import { logger, snooze } from './utils/utils'
+
+const client = createClient()
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/coinrankEngine.ts:10]: Makes sense. done

File: src/coinrankEngine.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+client.connect().catch(e => logger('redis connect error: ', e))
+
+const PAGE_SIZE = 250
+const LOOP_DELAY = 45000
+
+export const coinrankEngine = async (): Promise<void> => {
+  await console.log('hello')
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/coinrankEngine.ts:17]: done

File: src/coinrankEngine.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  logger('Updating Coinrank Cache')
+  try {
+    const lastUpdate = new Date().toISOString()
+    const { uri } = config.providers.coingecko
+    let markets: any = []
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/coinrankEngine.ts:23]: Nice. Done

File: src/coinrankEngine.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      if (response.ok === false) {
+        const text = await response.text()
+        throw new Error(text)
+      }
+      const reply = await response.json()
+      try {
+        const marketsPage = asCoingeckoMarkets(reply)
+        markets = [...markets, ...marketsPage]
+      } catch (e) {
+        console.log('ugh')
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/coinrankEngine.ts:36]: Makes sense. done

File: src/exchangeRateRouter.ts (hunk 2 above)
Comment (paullinator) [reply src/exchangeRateRouter.ts:349]: ok. note there are other hardcoded `iso:USD` occurences in the file

File: src/exchangeRateRouter.ts (hunk 1 above)
Comment (paullinator) [reply src/exchangeRateRouter.ts:390]: I don't understand. All USD market are queried by the engine. The API endpoint only writes non-USD markets after doing the conversion. The converted value is returned to the user and saved in redis.

Comment (peachbits) [reply src/exchangeRateRouter.ts:392]: The data saved to redis is only a subset of the USD data (sliced at 378 and saved in line 392) and once it is saved subsequent requests for the same fiat code will no longer be null at line 361 which skips any conversion functionality.

Instead, take markets in line 377 and convert all of them to the requested fiat -> save to redis -> slice and return to user

**Review by peachbits — CHANGES_REQUESTED** [review]

File: src/exchangeRateRouter.ts (hunk 1 above)
Comment (paullinator) [reply src/exchangeRateRouter.ts:392]: understood. makes sense.

File: src/indexEngines.ts [hunk 7]
```diff
@@ -10,8 +10,8 @@ import { logger } from './utils/utils'
 // Initialize DB
 async function initDb(): Promise<void> {
   await setupDatabase(config.couchUri, ratesDbSetup)
-  ratesEngine().catch(e => logger('ratesEnginee failure', e))
-  uidEngine().catch(e => logger('uidEnginee failure', e))
+  // ratesEngine().catch(e => logger('ratesEnginee failure', e))
```
Comment (peachbits) [inline src/indexEngines.ts:13]: turd here

File: src/indexEngines.ts [hunk 8]
```diff
@@ -10,8 +10,8 @@ import { logger } from './utils/utils'
...
+  // uidEngine().catch(e => logger('uidEnginee failure', e))
```
Comment (peachbits) [inline src/indexEngines.ts:14]: another turd

**Review by peachbits — CHANGES_REQUESTED** [review]

File: src/indexEngines.ts (hunk 7 above)
Comment (paullinator) [reply src/indexEngines.ts:13]: damn. the whole file change is a turd. how'd that sneak in there. will remove

**Review by peachbits — APPROVED** [review]

**Pull request merged**